### PR TITLE
chore: optimize building repository

### PR DESCRIPTION
**Pull request by AllanFly120**
* Switching over to yarn workspace to optimize the repo's bootstrap performance. The full repo bootstrap time is reduced to ~70 seconds.
* The `test` scripts now only compile and test the packages that:
    - changed from the last tag
    -  have direct or indirect dependency over changed packages
    - have direct or inderect dependents that are changed packages
* Add a 'test-all' script to compile/test all the packages

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

**Comment by AllanFly120** [comment]
This is ready to be reviewed. I'm still using `npm run script` because `yarn run script` currently doesn't seem to conform `pre` and `post` script hook. I will look into this later. Now the scripts all work fine with `npm` command.

**Comment by trivikr** [comment]
Let.s use yarn for everything as follows:
* Replace `npm run` with `yarn` in all package.json files
* Replace npm commands with yarn in `CONTRIBUTING.md` and `.travis.yml`:
  * replace `npm install` with `yarn`
  * remove `npm run bootstrap` (as yarn now does bootstrapping)
  * replace `npm test` with `yarn test`
* Remove `yarn.lock` from `.gitignore` so that we commit lock files and lock our dependencies

**Comment by AllanFly120** [comment]
@trivikr Updated all the scripts to use yarn already #327 
Some generated clients still have local `npm` scripts. Will update them later.

**Comment by trivikr** [comment]
I ran the following commands on 2473c5b9fd05ec563a8ac5a54edd326322dff87f in addition to removing `yarn.lock` from `.gitignore` on [m4.2xlarge](https://aws.amazon.com/ec2/instance-types/):

yarn --version: 1.17.3
* `git clean -dfx`
* `yarn`: 44.43s ([yarn.log](https://github.com/aws/aws-sdk-js-v3/files/3527495/yarn.log))
* `yarn test`: 1230.97s ([yarn-test.log](https://github.com/aws/aws-sdk-js-v3/files/3527497/yarn-test.log))

File: .travis.yml [hunk 1]
```diff
@@ -1,10 +1,12 @@
...
+before_install:
+  - curl -o- -L https://yarnpkg.com/install.sh | bash -s -- --version 1.17.3
+  - export PATH="$HOME/.yarn/bin:$PATH"
 install:
-  - npm install
-  - travis_wait 180 npm run bootstrap
+  - yarn
 script:
-  - travis_wait 180 npm run test
+  - yarn run test
```
Comment (trivikr) [inline .travis.yml]: `yarn test` would suffice ([doc](https://yarnpkg.com/lang/en/docs/cli/test/))
```suggestion
  - yarn test
```

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -67,10 +65,9 @@ Please be aware of the following notes prior to opening a pull request:
    [conventional commits][conventional commits]. Title must begin with `feat(module): title`,
    `fix(module): title`, `docs(module): title`, `test(module): title`, `chore(module): title`.
    Title should be lowercase and not period at the end of it. If the commit includes
-   a breaking change, the commit message must end with a single paragraph: `BREAKING 
-   CHANGE: a description of what broke` 
+   a breaking change, the commit message must end with a single paragraph: `BREAKING CHANGE: a description of what broke`
 
-5. After getting ready to open a pull request, make sure to run the `npm run update-clients`
+5. After getting ready to open a pull request, make sure to run the `yarn run update-clients`
```
Comment (trivikr) [inline CONTRIBUTING.md]:
```suggestion
5. After getting ready to open a pull request, make sure to run the `yarn update-clients`
```

File: buildspec.yml [hunk 3]
```diff
@@ -4,11 +4,11 @@ phases:
...
-      - echo Installing dependencies
-      - npm install && npm run bootstrap
+      - curl -o- -L https://yarnpkg.com/install.sh | bash -s -- --version 1.17.3
+      - export PATH="$HOME/.yarn/bin:$PATH"
+      - yarn
   build:
     commands:
       - echo Executing unit tests
-      - npm test
+      - yarn run test
```
Comment (trivikr) [inline buildspec.yml]:
```suggestion
      - yarn test
```

File: package.json [hunk 4]
```diff
@@ -5,13 +5,16 @@
   "description": "AWS SDK for JavaScript from the future",
   "main": "index.js",
   "scripts": {
-    "bootstrap": "lerna bootstrap",
-    "clean": "npm run clear-build-cache && lerna clean",
+    "bootstrap": "yarn install",
```
Comment (trivikr) [inline package.json]:
```suggestion
    "bootstrap": "yarn",
```

File: package.json [hunk 5]
```diff
@@ -5,13 +5,16 @@
...
+    "clean": "yarn run clear-build-cache && yarn run clear-build-info && lerna clean",
```
Comment (trivikr) [inline package.json]:
```suggestion
    "clean": "yarn clear-build-cache && yarn clear-build-info && lerna clean",
```

File: package.json [hunk 6]
```diff
@@ -5,13 +5,16 @@
...
     "clear-build-cache": "rimraf ./packages/*/build/* ./clients/*/*/build/*",
+    "clear-build-info": "rimraf ./packages/*/*.tsbuildinfo ./clients/*/*/*.tsbuildinfo",
     "copy-models": "node ./scripts/copyModels.js",
-    "update-clients": "node ./scripts/rebuildClients.js",
-    "pretest": "lerna run pretest",
-    "test": "jest --coverage"
+    "update-clients": "node ./packages/package-generator/build/cli.js import-all --matching '/Users/zheallan/workspace/aws-sdk-js-v3/models/*/*/service-2.json'",
```
Comment (trivikr) [inline package.json]: zheallan? 🤔

**Pull request merged**